File: Aditya Vardhan Singh/Other Problems/Day 12/19 Remove Nth Node From End of List/Java/19 Remove Nth Node from End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode header;
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int m = size(head) - n;
        if (head == null) {
            return head;
        }
        if (m == 0) {
            head = head.next;
            return head;
        }
        ListNode currNode = head;
        ListNode nextNode = head.next;
        for (int i = 1; i < m; i++) {
            currNode = currNode.next;
            nextNode = nextNode.next;
        }
        currNode.next = nextNode.next;
        return head;
    }

    public int size(ListNode head) {
        ListNode currNode = head;
        int size = 0;
        while (currNode != null) {
            size++;
            currNode = currNode.next;
        }
        return size;
    }
}

File: Aditya Vardhan Singh/Other Problems/Day 12/19 Remove Nth Node From End of List/Python/19 Remove Nth Node from End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        size = 0
        temp = head
        while temp:
            size += 1
            temp = temp.next
        
        m = size - n + 1

        if m == 1:
            head = head.next
        else:
            count = m - 2
            temp = head
            while count:
                temp = temp.next
                count -= 1
            temp.next = temp.next.next
        
        return head
